tests/bluetooth/bsim_bt/bsim_test_l2cap/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,563 @@
+			LOG_DBG("Disconnecting channel: %d)", channels[i].chan_id);
+			err = bt_l2cap_chan_disconnect(&channels[i].le.chan);
+			if (err) {
+				LOG_DBG("can't disconnnect channel (err: %d)", err);
+			}
+			channels[i].in_use = false;
+		}
+	}
+}
+
+static int accept(struct bt_conn *conn, struct bt_l2cap_chan **l2cap_chan)
+{
+	struct channel *chan;
+
+	chan = get_free_channel();
+	if (!chan) {
+		return -ENOMEM;
+	}
+
+	chan->le.chan.ops = &l2cap_ops;
+	chan->le.tx.mtu = DATA_MTU;
+	chan->le.rx.mtu = DATA_MTU;
+
+	*l2cap_chan = &chan->le.chan;
+
+	return 0;
+}
+
+static struct bt_l2cap_server *get_free_server(void)
+{
+	for (int i = 0U; i < SERVERS; i++) {
+		if (servers[i].psm) {
+			continue;
+		}
+
+		return &servers[i];
+	}
+
+	return NULL;
+}
+
+static void register_l2cap_server(void)
+{
+	struct bt_l2cap_server *server;
+
+	server = get_free_server();
+	if (!server) {
+		FAIL("Failed to get free server");
+		return;
+	}
+
+	server->accept = accept;
+	server->psm = 0;
+
+	if (bt_l2cap_server_register(server) < 0) {
+		FAIL("Failed to get free server");
+		return;
+	}
+
+	LOG_DBG("L2CAP server registered, PSM:0x%X", server->psm);
+}
+
+static void connected(struct bt_conn *conn, uint8_t conn_err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (conn_err) {
+		FAIL("Failed to connect to %s (%u)", addr, conn_err);
+		bt_conn_unref(default_conn);
+		default_conn = NULL;
+		return;
+	}
+
+	default_conn = bt_conn_ref(conn);
+	LOG_DBG("%s", addr);
+
+	SET_FLAG(is_connected);
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	LOG_DBG("%s (reason 0x%02x)", addr, reason);
+
+	if (default_conn != conn) {
+		FAIL("Conn mismatch disconnect %s %s)", default_conn, conn);
+		return;
+	}
+
+	bt_conn_unref(default_conn);
+	default_conn = NULL;
+	UNSET_FLAG(is_connected);
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+static void send_sdu(int itteration, int chan_idx, int bytes);
+
+void send_sdu_chan_worker(struct k_work *item)
+{
+	struct channel  *ch =
+		CONTAINER_OF(item, struct channel, work);
+
+	send_sdu(ch->itteration, ch->chan_id, ch->bytes_to_send);
+}
+
+static void send_sdu(int itteration, int chan_idx, int bytes)
+{
+	struct bt_l2cap_chan *chan = &channels[chan_idx].le.chan;
+
+	*((int *)channels[chan_idx].payload) = itteration; /*First 4 bytes in sent payload is itteration count*/
+	if (channels[chan_idx].buf != 0) {
+		LOG_ERR("Buf should have been dealocated by now");
+	}
+	struct net_buf *buf = net_buf_alloc(&tx_data_pool, K_NO_WAIT);
+
+	if (buf == 0) {
+		LOG_ERR("Failed to get buff on ch %i, iteration %i should never happen", chan_idx, chan_idx);
+	}
+	channels[chan_idx].buf = buf;
+	net_buf_reserve(buf, BT_L2CAP_CHAN_SEND_RESERVE);
+	net_buf_add_mem(buf, channels[chan_idx].payload, bytes);
```

Comment:
Suggest to use `net_buf_push_le32` to push the iteration before adding the payload.